<a href="https://colab.research.google.com/github/gt-cse-6040/big-o/blob/main/time_complexity.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Running Time of Algorithms and big-O notation

Did you ever think about how much time it takes the computer to run your program?
Consider a short function that calculates Fibonacci numbers.
In the Fibonnaci sequence every element equals to the sum of the two previous elements:

$$0,1,1,2,3,5,8,13,21...$$

The function `fibonacci` is given the index of the sequence and returns the corresponding value for this index.
for example, `fibonnaci(7)`->13, `fibonnaci(2)`->1, `fibonnaci(0)`->0

(By the way, if you don't understand the implementation of this program, do not worry! You can regard it as a black box. This notebook focuses on the running time of algorithms rather than on programming skills)

In [ ]:
def fibonacci(n):
    if (n==0):
        return 0, 1
    elif n==1 or n==2:
        return 1, 1
    else:
        fib2 = fibonacci(n-2)
        fib1 = fibonacci(n-1)
        return fib2[0] + fib1[0], fib2[1]+fib1[1]+1

In [ ]:
f7 = fibonacci(7)[0]
f2 = fibonacci(2)[0]
f20 = fibonacci(20)[0]

print("the 7th number of Fibonnaci is: ", f7)
print("the 2nd number of Fibonnaci is: ", f2)
print("the 20th number of Fibonnaci is: ", f20)



the 7th number of Fibonnaci is:  13
the 2nd number of Fibonnaci is:  1
the 20th number of Fibonnaci is:  6765


It ran pretty fast right? Now try to run this Fibonacci number with 30 as a parameter.

In [ ]:
fibonacci(30)[0]

832040

It took more time, but still pretty fast. What about 35? 100? (the next cell can take a while).

In [ ]:
fibonacci(35)[0]

9227465

Actually, do not try to run `fibonacci(100)`, unless you want to wait in front of this notebook until your grandchildren have their own grandchildren.

By the end of this notebook, you will understand exactly why that's the case.

Let's try now to understand how we can think about a running time of a program. 
consider the following program: program(a)

In [ ]:
## program(a)
list_of_numbers = [1,40,3,-5,2,6,10]
sum = 0
for i in list_of_numbers:
    x = i*2
    sum += x
print ("sum is: ", sum)
    
    
    

sum is:  114


Program(a) is given a list of numbers, computes the multiplication by 2 of each number, sums it and prints that sum.
Only for simplicity purposes, let's say that each line in this program is one basic computer operation

`list_of_numbers = [1,40,3,-5,2,6,10]` is 1 computer operation

`sum = 0` is 1 computer operation

`x = i*2` is 1 computer operation 


And so on...

So how many computer operations did this program execute?

2 at the begining, 14 in the for loop (2 in each loop $\cdot$ *7 loops*), and 1 at the end.

*2+14+1 = 17* $\rightarrow$ 17 computer operations.

What about program(b), the same program but different size of list:

In [ ]:
## program(b)
list_of_numbers = [1,40,3,-5,2,6,10,2,3,4,5,6,7,8,9,10]
sum = 0
for i in list_of_numbers:
    x = i*2
    sum += x
print ("sum is: ", sum)


sum is:  222


There are 2 at the beginning, 32 in the for loop (2 in each loop for *16 loops*), and 1 at the end. This makes a total of 35 computer operations.

The number of operations the computer executed has grown, and that is because the length of the list is bigger and as a consequence more loop iterations were conducted. 

Hence we can conclude that the **running time of the program depends on the size of the data structure** the program iterates over. In other words, the **running time is a function of the length of the data structure.**

In our program, if we denote the function $f$ to be the number of operations performed and  $n$ to be the length of the data structure then: $$f(n)= 2 + 2n +1  =   2n + 3$$ Notice that the running time here is linear in the length of the data structure.

In practice we will not count basic operations in every program, it is not that simple (1 line of code can constitute a lot of basic computer operations).

**Instead we will upper bound the running time (i.e. $f(n)$) with a known function.**

### Big-O notation

First, the mathematical definition:
$ \text{We say that } f(n) \text{ is } O(g(n)) \text{ when there is a constant } c \text{ such that } f(n) \le cg(n) \text{ for big values of n.}$

This means that $f(n)$ is $O(g(n))$, i.e. $f(n)$ is smaller than a some constant multiple of $g(n)$. 

For example, the function $f(n)=2n+3$ is $O(n)$; i.e. $g(n)=n$ because if c=10, for instance, $f(n)=2n+3\le10n$ when n is big.

Notice that $2n+3$ is also $O(n^2)$ and $O(n^{10})$ but we will try to find the slowest growing function possible.

So program(a) and program(b), that we ran before, have a *time complexity* of $O(n)$ !! And notice that n is the size of the list.

More examples: 
$$ 3n^2 = O(n^2) \text{;    for } c=5\text{; }\;3n^2 \le 5n^2 \text{ (for big n)}$$
$$ n^3 +5n^2 +20 = O(n^3) \text{;   for } c=100;\;n^3 +5n^2 +20 \le 100n^2 \text{ (for big n)}$$
$$ n - \sqrt{n} = O(n) \text{;    for } c=20\text{; } n-\sqrt{n} \le 20n \text{ (for big n)}$$
$$ 100 = O(1) \text{;    for } c=101\text{; } 100 \le 101\cdot1 \text{ (for big n)}$$
$$ 2^n = O(2^n) $$
$$ 3^n = O(3^n) $$
$$ 5\cdot log(n) + 100 = O(log(n))$$
$$ \sum_{i=1}^n i = 1+2+3+4+5+...+n = O(n^2)$$





**In general, we are not going to stick to the mathematical rules, instead we will try to understand how many times the program iterates our data set (without necessarily know how many operations have been done during each iteration), and we will choose the $O(\cdot)$ accordingly.** Let's go through some examples.

### Example no.1 - Searching in a sorted list 

Let's generate a sorted list: (it may take few moments)


In [ ]:
## randomize a integer list with length ~ n, and sort it.
import random
n = 10**7
lst = []
for i in range(n):
    x = random.randint(0,10**8)
    lst.append(x)
lst.append(12345678) ## inserting a  certain number to search
lst.sort()
print ("20 first elements: ", lst[:20])
print ("20 last elements:", lst[-20:])

20 first elements:  [3, 8, 12, 35, 54, 59, 67, 70, 75, 80, 84, 90, 99, 107, 110, 115, 123, 131, 150, 153]
20 last elements: [99999787, 99999787, 99999797, 99999808, 99999817, 99999820, 99999841, 99999843, 99999871, 99999874, 99999881, 99999881, 99999903, 99999906, 99999909, 99999911, 99999916, 99999924, 99999948, 99999951]


And let's say we want to find out if the number 12345678 is on the list or not.  
As a solution, we can just traverse the list from the beginning to the end and look for this element: (The answer should be true)

In [ ]:
def naive_search(lst, num):
    iterations=0
    for x in lst:
        iterations+=1
        if x==num:
            return True, iterations
    return False, iterations

In [ ]:
import time
t1 = time.perf_counter() ## assign into t1 the current time
flag, iterations = naive_search(lst, 12345678)
t2 = time.perf_counter()## assign into t2 the current time
if flag:
    print ("The number exists in the list")
else:
    print ("The number does not exist in the list")
print("# of iterations: ",iterations)
print("time: ", (t2-t1), "seconds") ## the time delta between t1-->t2


The number exists in the list
# of iterations:  1234891
time:  0.20563300399953732 seconds


What will be the number of iterations if the element we were looking for is one of the first elements? What if it is one of the last elements? What will be the number of iterations if the element is not part of the list?


**When analyzing *time complexity* of an algorithm we always consider the worst-case scenario.**
In the solution above, the worst-case scenario is if the element does not exist in the list, in this case, we will iterate over the whole list before ending the algorithms. 


Hence, in the worst case, we will have $n=10^7$ iterations, and that means the *time complexity* is $O(n)$, linear in the length of the data set.

Let's look for a number that does not exist:

In [ ]:
t1 = time.perf_counter() ## assign into t1 the current time
flag, iterations = naive_search(lst, 999999999999999)
t2 = time.perf_counter()## assign into t2 the current time
if flag:
    print ("The number exists in the list")
else:
    print ("The number does not exist in the list")
print("# of iterations: ",iterations)
print("time: ", (t2-t1), "seconds") ## the time delta between t1-->t2

The number does not exist in the list
# of iterations:  10000001
time:  1.6263801040004182 seconds


We are going now to improve our naive search, with a more sophisticated algorithm for searching in a sorted list, the Binary Search.
We will take advantage of the fact the list is sorted.
We try to find a number x, we will first go to the middle of the list and now we have 3 options:

1. the number is exactly in the middle, so we found him and we are done. 
2. the number is smaller than the middle, which means it can be found only in the first half of the list
3. the number is greater than the middle, which means it can be found only in the second half of the list.

if we are in step 2 or 3, we will redo the algorithm but only to the relevant half until we find (or don't find) the number. **Every iteration we cross out half of the list in our search! and that brings huge improvement to running time.**
The picture below is for illustration:
<img src="https://github.com/gt-cse-6040/big-o/blob/main/binary_search.png?raw=1">

In [ ]:
def binary_search(arr,x):
    l=0
    r=len(arr)-1
    iterations = 0
    while l <= r:
        iterations+=1
        mid = l + (r - l) // 2;   
        if arr[mid] == x: ## Check if x is in the middle
            return True, iterations
        elif arr[mid] < x:## If x is greater, ignore left half
            l = mid + 1
        else:             ## If x is smaller, ignore right half
            r = mid - 1
    return False, iterations  ## If we reach here, then the element is not in the list

Let's try this algorithm for the numbers we used in the naive search. 
First: 12345678 
Second, for the worst case scenario: 999999999999999

In [ ]:
t1 = time.perf_counter() ## assign into t1 the current time
flag, iterations = binary_search(lst, 12345678)
t2 = time.perf_counter()## assign into t2 the current time
if flag:
    print ("The number exists in the list")
else:
    print ("The number does not exist in the list")
print("# of iterations: ",iterations)
print("time: ", (t2-t1), "seconds") ## the time delta between t1-->t2

The number exists in the list
# of iterations:  24
time:  7.579199882457033e-05 seconds


In [ ]:
t1 = time.perf_counter() ## assign into t1 the current time
flag, iterations = binary_search(lst, 999999999999999)
t2 = time.perf_counter()## assign into t2 the current time
if flag:
    print ("The number exists in the list")
else:
    print ("The number does not exist in the list")
print("# of iterations: ",iterations)
print("time: ", (t2-t1), "seconds") ## the time delta between t1-->t2

The number does not exist in the list
# of iterations:  24
time:  7.625299986102618e-05 seconds


Let's summarize:

| algorithm| # of iterations for $n=10^7$ | time (sec)|
| --- | --- | --- |
| naive_search | 10000001 | 1.85 |
| binary_search | 24 | 0.000092 |

(the times are going to be different on your computers, but the trend should be the same)

Amazing! isn't it? 
Take a moment to think about the time complexity here. 
Every iteration we halved the list, which means that the number of iterations is the number of times we can half the list until we reach a list with the length of 1.

Hence if the length of the list is $n$ the *time complexity* is $O(log(n))$.
For our list, $n=10^7 \text{ and hence the number of iterations should be } \le \lceil log_2(10^7) \rceil = 24$  

Look how math really impacts our algorithms!

### Example no.2 - Sorting a list 

One of the most common tasks we do with lists is sort them. First let's generate a random list in the length of $n=10^4$ but this time we will not tell python to sort it. We will do it by ourselves: (it may take few moments)

In [ ]:
## randomize a integer list with length ~ n, and sort it.
import random
n = 10**4
lst = []
for i in range(n):
    x = random.randint(0,10**8)
    lst.append(x)
print ("20 first elements: ", lst[:20],"\n")
print ("20 last elements:", lst[-20:],"\n")

20 first elements:  [51705196, 46673070, 65147542, 78795793, 16903248, 73316659, 62421884, 12612289, 76537946, 65872794, 8453071, 59261202, 92507716, 63341494, 65539385, 84641224, 8814088, 24597687, 65807755, 6093827] 

20 last elements: [31008242, 33845853, 97104031, 30508009, 41505054, 4755710, 12816912, 89731171, 68525287, 51045805, 41715630, 12926701, 77091391, 79746581, 88816993, 60762081, 26944656, 41127252, 74497896, 26135987] 



Now let's sort it, we will first use a naive sort method, called selection sort. In selection sort we look for the minimum element in the list (what is the *time complexity* of finding the minimum in a list?) and then put this element at the beginning of the list. Then, we do the same for ```list[1:]``` and then for ```list[2:]``` until the end:

![SegmentLocal](Selection-Sort-Animation.gif "segment")

In [ ]:
def selection_sort(arr):
    iterations=0
    for i in range(len(arr)):
        min_idx = i
        for j in range(i, len(arr)):
            iterations+=1
            if arr[min_idx] > arr[j]:
                min_idx = j     
        arr[i], arr[min_idx] = arr[min_idx], arr[i]
    return arr, iterations

In [ ]:
t1 = time.perf_counter() ## assign into t1 the current time
lst, iterations = selection_sort(lst)
t2 = time.perf_counter()## assign into t2 the current time
print ("20 first elements: ", lst[:20],"\n")
print ("20 last elements:", lst[-20:],"\n")
print("# of iterations: ",iterations)
print("time: ", (t2-t1), "seconds") ## the time delta between t1-->t2

20 first elements:  [26424, 32581, 33977, 45442, 49497, 58036, 59299, 67306, 76887, 78608, 93143, 93393, 100980, 108159, 131555, 132540, 149841, 152130, 167893, 184727] 

20 last elements: [99748629, 99750205, 99752350, 99761707, 99768513, 99788137, 99789084, 99798340, 99801087, 99842588, 99862569, 99869384, 99871811, 99918584, 99951564, 99957648, 99962816, 99987452, 99991401, 99999394] 

# of iterations:  50005000
time:  4.669906931998412 seconds


Done! But it took a while, right? And keep in mind this was only for a list of $n=10^4=10000$ elements. What would happen if it was a list of $n=10^6=1000000$ elements, It would become really time consuming to sort that list... 

Let's analyze the running time. Notice we have a nested loop here, for the first outer loop, we executed $n$ iterations to find the minimum of ```list[:]```, then for the next outer loop we executed $n-1$ iterations to find the minimum of ```list[1:]```, then we executed $n-2$ iterations to find the next minimum and so on.

Arithmetically speaking, the number of iterations in total for this algorithm is: $$\sum_{i=n}^1 i = n+(n-1)+(n-2)+(n-3)+...+1 = \frac{n^2}{2} + \frac{n}{2} = O(n^2)$$
 
This *time complexity* of $O(n^2)$ is completely doable for computers for $n \approx 10^5$, However, for larger $n$ than that, the performance of the average computer will start to degrade.

Now, Let's improve it. Did you notice that in example no.1 we created a list of $n=10^7$ and the built-in ```sort()``` function of python sorted this list quickly? For those sizes of list python does not use selection sort but another sorting algorithm. There are many sort algorithms that use a divide and conquer method. Here we will explore Merge Sort.

Let's generate again a list of random numbers

In [ ]:
## randomize a integer list with length ~ n, and sort it.
import random
n = 10**4
lst = []
for i in range(n):
    x = random.randint(0,10**8)
    lst.append(x)
print ("20 first elements: ", lst[:20],"\n")
print ("20 last elements:", lst[-20:],"\n")

20 first elements:  [61885140, 89309032, 54402896, 42285095, 18415595, 15858615, 37547666, 25168086, 62740852, 28280591, 82253556, 25812327, 39045387, 36548647, 20013121, 11459858, 88884165, 8563679, 56577748, 61685126] 

20 last elements: [79978785, 44687725, 67070849, 12795938, 28100649, 35458754, 7176517, 14300816, 29755547, 87897155, 8389200, 4778933, 87891392, 84447376, 58338225, 60346636, 90596133, 44086894, 27843594, 76164972] 




On a high level the algorithm has three steps, which are repeated recursively:
1. Split the list in half if possible, otherwise the list of length 1 is considered sorted and passed up.
2. Sort each half recursively using this algorithm.
3. Merge the two halves together by "popping" the smallest element from the top of each half until you exhaust one half.

The algorithm is implemented below. 
(If you do not understand the implementation, do not worry. It is not the subject of this notebook)

In [ ]:
def merge(a, b, counter, verbose=False):
    result = []
    x = a.copy()
    y = b.copy()
    counter += len(x) + len(y)
    while (len(x) > 0 and len(y) > 0):
        if x[0] < y[0]:
            result.append(x[0])
            x.pop(0)
        else:
            result.append(y[0])
            y.pop(0)
    result = result + x + y
    
    return result, counter

def merge_sort(l, counter=0):
    if len(l) == 1: return l, counter
    m = len(l) // 2
    a, c1 = merge_sort(l[:m], counter)
    b, c2 = merge_sort(l[m:], counter)
    counter += c1 + c2
    return merge(a, b, counter)

In [ ]:
t1 = time.perf_counter() ## assign into t1 the current time
lst, iterations = merge_sort(lst)
t2 = time.perf_counter()## assign into t2 the current time
print ("20 first elements: ", lst[:20],"\n")
print ("20 last elements:", lst[-20:],"\n")
print("# of iterations: ",iterations)
print("time: ", (t2-t1), "seconds") ## the time delta between t1-->t2

20 first elements:  [876, 9027, 9962, 14919, 21034, 32615, 46797, 48200, 127296, 161479, 164517, 175063, 175622, 175802, 199059, 201743, 215074, 216625, 220736, 223623] 

20 last elements: [99796353, 99801667, 99806844, 99822598, 99830906, 99834176, 99876114, 99879269, 99880545, 99884876, 99888022, 99889721, 99893275, 99916674, 99920882, 99928986, 99954145, 99959165, 99971926, 99982464] 

# of iterations:  133616
time:  0.059926264999376144 seconds


Before we analyze, let's summarize:

| algorithm| # of iterations for $n=10^4$ | time (sec)|
| --- | --- | --- |
| selection_sort | 50005000 |  5.624 |
| merge_sort | 133616 | 0.0494 |

(the times are going to be different on your computers, but the trend should be the same)

Now, let's do some analysis:

First, let's look at the `merge` function. Specifically the inner portion of the `while` loop.
```
if x[0] < y[0]:
    result.append(x[0])
    x.pop(0)
else:
    result.append(y[0])
    y.pop(0)
```
This code has a conditional check, a list append, and a list pop. None of these depend on the length of the input lists (`a` and `b`), so the inner portion of the loop is $O(1)$.

The next question is "how many times does the $O(1)$ loop execute?" Well, it removes exactly one element every time through until one list is exhausted, and the worst case is that the largest two elements will be in separate lists. This means that we could iterate up to `len(a) + len(b)` times. Even if we aren't in the worst case we still have to add the rest of the non-empty list onto the result after the loop finishes. The operations that we save by getting lucky are made up immediately afterwards. If we define $k$ as `len(a) + len(b)`, then each `merge` step is $O(k)$.

Now let's look at the `merge_sort` function. We'll start with the return statement. 
```
return merge(a, b, counter)
```
This is just a call to `merge` and `a` and `b` are two halves of the list we're sorting (length `n`). So this step is $O(n)$. 

Now there's just the matter of the two recursive calls to `merge_sort`. Let's start by defining a function $T(n)$ that is the run time of a call to `merge_sort` sorting a list of length $n$. Note that each recursive call is working on a list that is **half** of the length of the original input, i.e. $n/2$
$$T(n) = 2T(\lceil n/2 \rceil) + O(n); \; T(1) = 0$$
Now consider $f(n) = 2f( \frac{n}{2} ) + n;\; f(n) = 0 \text{ if } n \le 1$. Lets expand a few times.
$$f(n) = n + 2\frac{n}{2}  + 4\frac{n}{4} + 8f\left (\frac{n}{8} \right)$$
We can see a pattern. Each time we expand we add another $n$ to the sum. Let's look at the $k$-th expansion.
$$f(n) = nk + 2^kf(\frac{n}{2^k})$$
Remember that eventually, once $\frac{n}{2^k} \le 1$ the additional terms will become $0$. 
$$\frac{n}{2^k} \le 1\;\rightarrow\;n \le 2^k\;\rightarrow\; log_2n \le k$$
With this we can conclude that the sum will have no more than $log_2n$ terms. $T(n) = O(nlog_2n)$!




### Example no.3 - Constant complexity 

Consider the following program:

In [ ]:
lst = [5,3,9,11,12,13,14,15,16,17,18,44,-1]
x=lst[0]
y=lst[1]
z=lst[2]
x=x+1
y=y+y
z=z-z*z
print ((x*y*z)%(z+y+z))

-108


Consider the above code: although we have few commands, we do not iterate the data set. We do a constant number of computer operations that do not depend on the length of the data structure. 

In this case we say that this program is $O(1)$. This is the best *time complexity* algorithm we can have. 

We will immediately see how we can improve very high *time complexity* to be $O(1)$. the following program: 

### Back to Fibonnaci numbers

Reminder that at the beginning of the notebook we created a function that calculates Fibonacci numbers

In [ ]:
def fibonacci(n):
    if (n==0):
        return 0, 1
    elif n==1 or n==2:
        return 1, 1
    else:
        fib2 = fibonacci(n-2)
        fib1 = fibonacci(n-1)
        return fib2[0] + fib1[0], fib2[1]+fib1[1]+1

What is the "length of the data structure" in this case? This function actually calculates the whole Fibonacci numbers sequence until the $n$ index and returns this $n's$ value. So in this case length of the data structure is $n$ and the running time is a function of $n$. 

Let's see how many iterations we do in the following: (it may take a while)

In [ ]:
idx = [1,7,20,30,35]
iters = []
times = []
for i in [1,7,20,30,35]:
    t1 = time.perf_counter() ## assign into t1 the current time
    iters.append(fibonacci(i)[1])
    t2 = time.perf_counter()## assign into t2 the current time
    times.append(t2-t1)
    
for i,x in enumerate(idx):
    print("for n=",x)
    print ("# of iterations: ",iters[i])
    print ("time: ",times[i],"seconds")
    print ("\n\n")

for n= 1
# of iterations:  1
time:  1.8549981177784503e-06 seconds



for n= 7
# of iterations:  25
time:  7.081998774083331e-06 seconds



for n= 20
# of iterations:  13529
time:  0.002610902000014903 seconds



for n= 30
# of iterations:  1664079
time:  0.3201591910001298 seconds



for n= 35
# of iterations:  18454929
time:  3.5393053199986753 seconds





See how quickly, for relatively small $n$, we climb up in the # of iterations and time.

Let's analyze the *time complexity* of this function. 
This function work as follows: for a given $n$, it calculates the Fibonacci number of $n-2$ and $n-1$ and sum them together, but to do those calculations we need to call again for this Fibonacci function. and each function call another Fibonacci function. It can be described in the following image: 
<img src="https://github.com/gt-cse-6040/big-o/blob/main/fibonacci_tree.png?raw=1">

In each node of the tree, we do a constant amount of operations. The number of nodes in the tree depends on $n$ and with some math we can show that the number of nodes is: $O(2^n)$. 

Since we do a constant amount of operations in each node the *time complexity* of this algorithm is also $O(2^n)$. Exponential. 
That is why even for small $n$, like $n=100$ it will take your computer $\approx 2^{100}$ operations which is $\approx 1 \text{ million years for a state of the art computer.}$   

Be very cautious when dealing with algorithms that can be exponential in $n$.

However, with number theory, we can show that:
$$fibonacci(n) = \frac{1}{\sqrt{5}}\cdot((\frac{1+\sqrt{5}}{2})^n - (\frac{1-\sqrt{5}}{2})^n)$$

...But we will spare you that ;)

We will write a function that implements the above formula:

In [ ]:
def fibonacci_better(n):
    return int((1/(5**0.5))*(((1+5**0.5)/2)**n-((1-5**0.5)/2)**n)//1)

In [ ]:
print ("fibonacci_better(7): ", fibonacci_better(7))
print ("fibonacci_better(20): ", fibonacci_better(20))
print ("fibonacci_better(35): ", fibonacci_better(35))

fibonacci_better(7):  13
fibonacci_better(20):  6765
fibonacci_better(35):  9227465


In [ ]:
print ("fibonacci_better(100): ", fibonacci_better(100))

fibonacci_better(100):  354224848179263111168


And this calculates Fibonacci numbers in $O(1)$ !!

### Conclusion

Keep in mind that the computer has limits, even when it comes to computational time. when you run your programs, especially when it deals with large data sets, be aware of the efficiency of the algorithm. As a rule of thumb, it is better to run algorithms that are $O(n \cdot log(n))$ or lower. Otherwise, it can take a lot of time to complete the algorithms.  
<img src="https://github.com/gt-cse-6040/big-o/blob/main/bigO_chart.png?raw=1">

### Extra

What is the *time complexity* of this permutation algorithm?

In [ ]:
def permutation(list_of_char, left, right, results):
    if left==right:
        results.append(list_of_char)
    else:
        for i in range(left,right+1):
            list_of_char[left], list_of_char[i] = list_of_char[i], list_of_char[left]
            permutation(list_of_char, left+1, right, results)
            list_of_char[left], list_of_char[i] = list_of_char[i], list_of_char[left] 
list_of_char = list(['a','b','c','d','1','!','*'])
perms = []

permutation(list_of_char, 0, len(list_of_char)-1, perms)
len(perms)

5040